File: tests/python/mathrobo_parity.py
#!/usr/bin/env python3

import json
import sys
from pathlib import Path

import numpy as np


REPO_ROOT = Path(__file__).resolve().parents[2]
MATHROBO_ROOT = REPO_ROOT / "vendor" / "MathRobo"
sys.path.insert(0, str(MATHROBO_ROOT))

import mathrobo as mr  # noqa: E402
from mathrobo.lie.se3 import SE3  # noqa: E402
from mathrobo.lie.so3 import SO3  # noqa: E402


def array(value):
    return np.asarray(value, dtype=float).tolist()


def main() -> None:
    so3_vector = np.array([0.2, -0.1, 0.3], dtype=float)
    so3_scale = 0.75
    so3_left = np.array([0.4, -0.2, 0.1], dtype=float)
    so3_right = np.array([-0.3, 0.5, 0.2], dtype=float)
    quaternion = np.array([0.9238795325, 0.0, 0.3826834324, 0.0], dtype=float)

    se3_twist = np.array([0.2, -0.1, 0.3, 1.0, -2.0, 0.5], dtype=float)
    se3_scale = 0.6
    se3_left_rot = SO3.exp(np.array([0.1, 0.2, -0.15], dtype=float))
    se3_right_rot = SO3.exp(np.array([-0.05, 0.3, 0.25], dtype=float))
    se3_left_pos = np.array([0.4, -0.8, 1.2], dtype=float)
    se3_right_pos = np.array([-1.0, 0.25, 0.75], dtype=float)
    se3_point = np.array([0.2, -0.4, 1.5], dtype=float)

    so3_left_obj = SO3(SO3.exp(so3_left))
    so3_right_obj = SO3(SO3.exp(so3_right))
    se3_left_obj = SE3(se3_left_rot, se3_left_pos)
    se3_right_obj = SE3(se3_right_rot, se3_right_pos)

    cmtm_so3_derivatives = np.array(
        [[0.1, -0.2, 0.3], [0.05, 0.01, -0.04]],
        dtype=float,
    )
    cmtm_so3_right_rotvec = np.array([-0.2, 0.15, 0.35], dtype=float)
    cmtm_so3_right_derivatives = np.array(
        [[-0.02, 0.04, 0.01], [0.03, -0.01, 0.02]],
        dtype=float,
    )
    cmtm_so3 = SO3(SO3.exp(np.array([0.25, -0.35, 0.1], dtype=float)))
    cmtm_so3_right = SO3(SO3.exp(cmtm_so3_right_rotvec))
    cmtm_so3_obj = mr.CMTM[SO3](cmtm_so3, cmtm_so3_derivatives)
    cmtm_so3_other = mr.CMTM[SO3](cmtm_so3_right, cmtm_so3_right_derivatives)

    cmtm_se3_rotvec = np.array([0.15, -0.2, 0.05], dtype=float)
    cmtm_se3_translation = np.array([0.4, -0.6, 0.8], dtype=float)
    cmtm_se3_derivatives = np.array(
        [[0.1, -0.05, 0.03, 1.0, -0.4, 0.2], [0.02, 0.01, -0.04, -0.3, 0.6, 0.5]],
        dtype=float,
    )
    cmtm_se3_right_rotvec = np.array([-0.05, 0.18, 0.22], dtype=float)
    cmtm_se3_right_translation = np.array([-0.7, 0.2, 0.1], dtype=float)
    cmtm_se3_right_derivatives = np.array(
        [[-0.03, 0.06, 0.02, 0.2, 0.1, -0.5], [0.04, -0.02, 0.01, -0.1, 0.3, 0.25]],
        dtype=float,
    )
    cmtm_se3 = SE3(SO3.exp(cmtm_se3_rotvec), cmtm_se3_translation)
    cmtm_se3_right = SE3(SO3.exp(cmtm_se3_right_rotvec), cmtm_se3_right_translation)
    cmtm_se3_obj = mr.CMTM[SE3](cmtm_se3, cmtm_se3_derivatives)
    cmtm_se3_other = mr.CMTM[SE3](cmtm_se3_right, cmtm_se3_right_derivatives)

    payload = {
        "so3": {
            "vector": array(so3_vector),
            "scale": so3_scale,
            "hat": array(SO3.hat(so3_vector)),
            "vee": array(SO3.vee(SO3.hat(so3_vector))),
            "exp": array(SO3.exp(so3_vector, so3_scale)),
            "compose_left": array(so3_left),
            "compose_right": array(so3_right),
            "compose": array((so3_left_obj @ so3_right_obj).mat()),
            "quaternion": array(quaternion),
            "quaternion_matrix": array(SO3.set_quaternion(quaternion).mat()),
        },
        "se3": {
            "twist": array(se3_twist),
            "scale": se3_scale,
            "hat": array(SE3.hat(se3_twist)),
            "vee": array(SE3.vee(SE3.hat(se3_twist))),
            "exp": array(SE3.exp(se3_twist, se3_scale)),
            "left_rotvec": array(np.array([0.1, 0.2, -0.15], dtype=float)),
            "left_translation": array(se3_left_pos),
            "right_rotvec": array(np.array([-0.05, 0.3, 0.25], dtype=float)),
            "right_translation": array(se3_right_pos),
            "compose": array((se3_left_obj @ se3_right_obj).mat()),
            "inverse": array(se3_left_obj.inv().mat()),
            "adjoint": array(se3_left_obj.mat_adj()),
            "point": array(se3_point),
            "apply": array(se3_left_obj @ se3_point),
        },
        "cmtm_so3": {
            "rotation_vector": array(np.array([0.25, -0.35, 0.1], dtype=float)),
            "derivatives": array(cmtm_so3_derivatives),
            "mat": array(cmtm_so3_obj.mat()),
            "mat_adj": array(cmtm_so3_obj.mat_adj()),
            "mat_inv": array(cmtm_so3_obj.mat_inv()),
            "mat_inv_adj": array(cmtm_so3_obj.mat_inv_adj()),
            "set_mat_vecs": array(mr.CMTM.set_mat(SO3, cmtm_so3_obj.mat()).vecs()),
            "inverse_mat": array(cmtm_so3_obj.inv().mat()),
            "compose_right_rotation_vector": array(cmtm_so3_right_rotvec),
            "compose_right_derivatives": array(cmtm_so3_right_derivatives),
            "compose_mat": array((cmtm_so3_obj @ cmtm_so3_other).mat()),
            "hat": array(mr.CMTM.hat(SO3, cmtm_so3_derivatives)),
            "hat_adj": array(mr.CMTM.hat_adj(SO3, cmtm_so3_derivatives)),
            "hat_commute_adj": array(mr.CMTM.hat_commute_adj(SO3, cmtm_so3_derivatives)),
            "vee": array(mr.CMTM.vee(SO3, mr.CMTM.hat(SO3, cmtm_so3_derivatives))),
            "vee_adj": array(mr.CMTM.vee_adj(SO3, mr.CMTM.hat_adj(SO3, cmtm_so3_derivatives))),
            "tangent_mat": array(cmtm_so3_obj.tangent_mat()),
            "tangent_mat_cm": array(cmtm_so3_obj.tangent_mat_cm()),
        },
        "cmtm_se3": {
            "rotation_vector": array(cmtm_se3_rotvec),
            "translation": array(cmtm_se3_translation),
            "derivatives": array(cmtm_se3_derivatives),
            "mat": array(cmtm_se3_obj.mat()),
            "mat_adj": array(cmtm_se3_obj.mat_adj()),
            "mat_inv": array(cmtm_se3_obj.mat_inv()),
            "mat_inv_adj": array(cmtm_se3_obj.mat_inv_adj()),
            "set_mat_vecs": array(mr.CMTM.set_mat(SE3, cmtm_se3_obj.mat()).vecs()),
            "inverse_mat": array(cmtm_se3_obj.inv().mat()),
            "compose_right_rotation_vector": array(cmtm_se3_right_rotvec),
            "compose_right_translation": array(cmtm_se3_right_translation),
            "compose_right_derivatives": array(cmtm_se3_right_derivatives),
            "compose_mat": array((cmtm_se3_obj @ cmtm_se3_other).mat()),
            "hat": array(mr.CMTM.hat(SE3, cmtm_se3_derivatives)),
            "hat_adj": array(mr.CMTM.hat_adj(SE3, cmtm_se3_derivatives)),
            "hat_commute_adj": array(mr.CMTM.hat_commute_adj(SE3, cmtm_se3_derivatives)),
            "vee": array(mr.CMTM.vee(SE3, mr.CMTM.hat(SE3, cmtm_se3_derivatives))),
            "vee_adj": array(mr.CMTM.vee_adj(SE3, mr.CMTM.hat_adj(SE3, cmtm_se3_derivatives))),
            "tangent_mat": array(cmtm_se3_obj.tangent_mat()),
            "tangent_mat_cm": array(cmtm_se3_obj.tangent_mat_cm()),
        },
    }

    json.dump(payload, sys.stdout)


if __name__ == "__main__":
    main()

File: tests/mathrobo_parity.rs
use std::path::PathBuf;
use std::process::Command;
use std::sync::OnceLock;

use mathroborust::{RustSe3, RustSo3};
use serde::Deserialize;

#[derive(Deserialize)]
struct ParityPayload {
    so3: So3Parity,
    se3: Se3Parity,
    cmtm_so3: CmtmSo3Parity,
    cmtm_se3: CmtmSe3Parity,
}

#[derive(Deserialize)]
struct So3Parity {
    vector: [f64; 3],
    scale: f64,
    hat: [[f64; 3]; 3],
    vee: [f64; 3],
    exp: [[f64; 3]; 3],
    compose_left: [f64; 3],
    compose_right: [f64; 3],
    compose: [[f64; 3]; 3],
    quaternion: [f64; 4],
    quaternion_matrix: [[f64; 3]; 3],
}

#[derive(Deserialize)]
struct Se3Parity {
    twist: [f64; 6],
    scale: f64,
    hat: [[f64; 4]; 4],
    vee: [f64; 6],
    exp: [[f64; 4]; 4],
    left_rotvec: [f64; 3],
    left_translation: [f64; 3],
    right_rotvec: [f64; 3],
    right_translation: [f64; 3],
    compose: [[f64; 4]; 4],
    inverse: [[f64; 4]; 4],
    adjoint: [[f64; 6]; 6],
    point: [f64; 3],
    apply: [f64; 3],
}

#[derive(Deserialize)]
struct CmtmSo3Parity {
    rotation_vector: [f64; 3],
    derivatives: Vec<[f64; 3]>,
    mat: Vec<Vec<f64>>,
    mat_adj: Vec<Vec<f64>>,
    mat_inv: Vec<Vec<f64>>,
    mat_inv_adj: Vec<Vec<f64>>,
    set_mat_vecs: Vec<[f64; 3]>,
    inverse_mat: Vec<Vec<f64>>,
    compose_right_rotation_vector: [f64; 3],
    compose_right_derivatives: Vec<[f64; 3]>,
    compose_mat: Vec<Vec<f64>>,
    hat: Vec<Vec<f64>>,
    hat_adj: Vec<Vec<f64>>,
    hat_commute_adj: Vec<Vec<f64>>,
    vee: Vec<[f64; 3]>,
    vee_adj: Vec<[f64; 3]>,
    tangent_mat: Vec<Vec<f64>>,
    tangent_mat_cm: Vec<Vec<f64>>,
}

#[derive(Deserialize)]
struct CmtmSe3Parity {
    rotation_vector: [f64; 3],
    translation: [f64; 3],
    derivatives: Vec<[f64; 6]>,
    mat: Vec<Vec<f64>>,
    mat_adj: Vec<Vec<f64>>,
    mat_inv: Vec<Vec<f64>>,
    mat_inv_adj: Vec<Vec<f64>>,
    set_mat_vecs: Vec<[f64; 6]>,
    inverse_mat: Vec<Vec<f64>>,
    compose_right_rotation_vector: [f64; 3],
    compose_right_translation: [f64; 3],
    compose_right_derivatives: Vec<[f64; 6]>,
    compose_mat: Vec<Vec<f64>>,
    hat: Vec<Vec<f64>>,
    hat_adj: Vec<Vec<f64>>,
    hat_commute_adj: Vec<Vec<f64>>,
    vee: Vec<[f64; 6]>,
    vee_adj: Vec<[f64; 6]>,
    tangent_mat: Vec<Vec<f64>>,
    tangent_mat_cm: Vec<Vec<f64>>,
}

fn repo_root() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
}

fn python_command() -> String {
    std::env::var("PYTHON").unwrap_or_else(|_| "python3".to_string())
}

fn load_parity_payload() -> ParityPayload {
    let script = repo_root().join("tests/python/mathrobo_parity.py");
    let output = Command::new(python_command())
        .arg(script)
        .current_dir(repo_root())
        .output()
        .expect("failed to run MathRobo parity script");

    assert!(
        output.status.success(),
        "MathRobo parity script failed.\nstdout:\n{}\nstderr:\n{}",
        String::from_utf8_lossy(&output.stdout),
        String::from_utf8_lossy(&output.stderr)
    );

    serde_json::from_slice(&output.stdout).expect("failed to parse MathRobo parity payload")
}

fn parity_payload() -> &'static ParityPayload {
    static PAYLOAD: OnceLock<ParityPayload> = OnceLock::new();
    PAYLOAD.get_or_init(load_parity_payload)
}

fn approx_eq_vec(actual: &[f64], expected: &[f64], tol: f64) {
    assert_eq!(actual.len(), expected.len());
    for (idx, (a, e)) in actual.iter().zip(expected.iter()).enumerate() {
        assert!(
            (a - e).abs() <= tol,
            "mismatch at index {idx}: expected {e}, got {a}"
        );
    }
}

fn approx_eq_matrix<const R: usize, const C: usize>(
    actual: &[[f64; C]; R],
    expected: &[[f64; C]; R],
    tol: f64,
) {
    for row in 0..R {
        for col in 0..C {
            let a = actual[row][col];
            let e = expected[row][col];
            assert!(
                (a - e).abs() <= tol,
                "mismatch at ({row},{col}): expected {e}, got {a}"
            );
        }
    }
}

fn approx_eq_matrix_dynamic(actual: &[Vec<f64>], expected: &[Vec<f64>], tol: f64) {
    assert_eq!(actual.len(), expected.len());
    for (row_idx, (actual_row, expected_row)) in actual.iter().zip(expected.iter()).enumerate() {
        assert_eq!(actual_row.len(), expected_row.len());
        for (col_idx, (a, e)) in actual_row.iter().zip(expected_row.iter()).enumerate() {
            assert!(
                (a - e).abs() <= tol,
                "mismatch at ({row_idx},{col_idx}): expected {e}, got {a}"
            );
        }
    }
}

fn approx_eq_vector_list<const DIM: usize>(
    actual: &[[f64; DIM]],
    expected: &[[f64; DIM]],
    tol: f64,
) {
    assert_eq!(actual.len(), expected.len());
    for (idx, (actual_vec, expected_vec)) in actual.iter().zip(expected.iter()).enumerate() {
        for component in 0..DIM {
            let a = actual_vec[component];
            let e = expected_vec[component];
            assert!(
                (a - e).abs() <= tol,
                "mismatch at vector {idx}, component {component}: expected {e}, got {a}"
            );
        }
    }
}

fn matrix6_to_array(matrix: nalgebra::SMatrix<f64, 6, 6>) -> [[f64; 6]; 6] {
    let mut out = [[0.0; 6]; 6];
    for row in 0..6 {
        for col in 0..6 {
            out[row][col] = matrix[(row, col)];
        }
    }
    out
}

fn dmatrix_to_vec(matrix: nalgebra::DMatrix<f64>) -> Vec<Vec<f64>> {
    (0..matrix.nrows())
        .map(|row| (0..matrix.ncols()).map(|col| matrix[(row, col)]).collect())
        .collect()
}

#[test]
fn so3_matches_mathrobo_reference() {
    let so3 = &parity_payload().so3;

    approx_eq_matrix(&RustSo3::hat(so3.vector), &so3.hat, 1e-12);
    approx_eq_vec(&RustSo3::vee(so3.hat), &so3.vee, 1e-12);
    approx_eq_matrix(&RustSo3::exp(so3.vector, Some(so3.scale)), &so3.exp, 1e-12);

    let left = RustSo3::from_rotation_vector(so3.compose_left);
    let right = RustSo3::from_rotation_vector(so3.compose_right);
    approx_eq_matrix(&left.compose(&right).to_matrix(), &so3.compose, 1e-12);

    let from_quaternion = RustSo3::from_quaternion(so3.quaternion);
    approx_eq_matrix(&from_quaternion.to_matrix(), &so3.quaternion_matrix, 1e-11);
}

#[test]
fn se3_matches_mathrobo_reference() {
    let se3 = &parity_payload().se3;

    approx_eq_matrix(&RustSe3::hat(se3.twist), &se3.hat, 1e-12);
    approx_eq_vec(&RustSe3::vee(se3.hat), &se3.vee, 1e-12);
    approx_eq_matrix(&RustSe3::exp(se3.twist, Some(se3.scale)), &se3.exp, 1e-12);

    let left = RustSe3::from_parts(
        RustSo3::from_rotation_vector(se3.left_rotvec),
        se3.left_translation,
    );
    let right = RustSe3::from_parts(
        RustSo3::from_rotation_vector(se3.right_rotvec),
        se3.right_translation,
    );

    approx_eq_matrix(&left.compose(&right).to_matrix(), &se3.compose, 1e-12);
    approx_eq_matrix(&left.inverse().to_matrix(), &se3.inverse, 1e-12);
    approx_eq_matrix(&matrix6_to_array(left.adjoint()), &se3.adjoint, 1e-12);
    approx_eq_vec(&left.apply(se3.point), &se3.apply, 1e-12);
}

#[test]
fn cmtm_so3_matches_mathrobo_reference() {
    let cmtm = &parity_payload().cmtm_so3;

    let left = mathroborust::RotationalCmtm::from_so3_with_derivatives(
        &RustSo3::from_rotation_vector(cmtm.rotation_vector),
        cmtm.derivatives.clone(),
    );

    approx_eq_matrix_dynamic(&dmatrix_to_vec(left.mat(None)), &cmtm.mat, 1e-10);
    approx_eq_matrix_dynamic(&dmatrix_to_vec(left.mat_adj(None)), &cmtm.mat_adj, 1e-10);
    approx_eq_matrix_dynamic(&dmatrix_to_vec(left.mat_inv(None)), &cmtm.mat_inv, 1e-10);
    approx_eq_matrix_dynamic(
        &dmatrix_to_vec(left.mat_inv_adj(None)),
        &cmtm.mat_inv_adj,
        1e-10,
    );

    let set_from_mat = mathroborust::RotationalCmtm::set_mat(&left.mat(None));
    approx_eq_vector_list(&set_from_mat.vecs(None), &cmtm.set_mat_vecs, 1e-10);

    approx_eq_matrix_dynamic(
        &dmatrix_to_vec(left.inverse().mat(None)),
        &cmtm.inverse_mat,
        1e-10,
    );

    let right = mathroborust::RotationalCmtm::from_so3_with_derivatives(
        &RustSo3::from_rotation_vector(cmtm.compose_right_rotation_vector),
        cmtm.compose_right_derivatives.clone(),
    );
    approx_eq_matrix_dynamic(
        &dmatrix_to_vec(left.compose(&right).mat(None)),
        &cmtm.compose_mat,
        1e-10,
    );

    approx_eq_matrix_dynamic(
        &dmatrix_to_vec(mathroborust::RotationalCmtm::hat(&cmtm.derivatives)),
        &cmtm.hat,
        1e-10,
    );
    approx_eq_matrix_dynamic(
        &dmatrix_to_vec(mathroborust::RotationalCmtm::hat_adj(&cmtm.derivatives)),
        &cmtm.hat_adj,
        1e-10,
    );
    approx_eq_matrix_dynamic(
        &dmatrix_to_vec(mathroborust::RotationalCmtm::hat_commute_adj(
            &cmtm.derivatives,
        )),
        &cmtm.hat_commute_adj,
        1e-10,
    );
    approx_eq_vector_list(
        &mathroborust::RotationalCmtm::vee(&mathroborust::RotationalCmtm::hat(&cmtm.derivatives)),
        &cmtm.vee,
        1e-10,
    );
    approx_eq_vector_list(
        &mathroborust::RotationalCmtm::vee_adj(&mathroborust::RotationalCmtm::hat_adj(
            &cmtm.derivatives,
        )),
        &cmtm.vee_adj,
        1e-10,
    );
    approx_eq_matrix_dynamic(
        &dmatrix_to_vec(left.tangent_mat(None)),
        &cmtm.tangent_mat,
        1e-10,
    );
    approx_eq_matrix_dynamic(
        &dmatrix_to_vec(left.tangent_mat_cm(None)),
        &cmtm.tangent_mat_cm,
        1e-10,
    );
}

#[test]
fn cmtm_se3_matches_mathrobo_reference() {
    let cmtm = &parity_payload().cmtm_se3;

    let left = mathroborust::RustCmtm::from_se3_with_derivatives(
        &RustSe3::from_parts(
            RustSo3::from_rotation_vector(cmtm.rotation_vector),
            cmtm.translation,
        ),
        cmtm.derivatives.clone(),
    );

    approx_eq_matrix_dynamic(&dmatrix_to_vec(left.mat(None)), &cmtm.mat, 1e-10);
    approx_eq_matrix_dynamic(&dmatrix_to_vec(left.mat_adj(None)), &cmtm.mat_adj, 1e-10);
    approx_eq_matrix_dynamic(&dmatrix_to_vec(left.mat_inv(None)), &cmtm.mat_inv, 1e-10);
    approx_eq_matrix_dynamic(
        &dmatrix_to_vec(left.mat_inv_adj(None)),
        &cmtm.mat_inv_adj,
        1e-10,
    );

    let set_from_mat = mathroborust::RustCmtm::set_mat(&left.mat(None));
    approx_eq_vector_list(&set_from_mat.vecs(None), &cmtm.set_mat_vecs, 1e-10);

    approx_eq_matrix_dynamic(
        &dmatrix_to_vec(left.inverse().mat(None)),
        &cmtm.inverse_mat,
        1e-10,
    );

    let right = mathroborust::RustCmtm::from_se3_with_derivatives(
        &RustSe3::from_parts(
            RustSo3::from_rotation_vector(cmtm.compose_right_rotation_vector),
            cmtm.compose_right_translation,
        ),
        cmtm.compose_right_derivatives.clone(),
    );
    approx_eq_matrix_dynamic(
        &dmatrix_to_vec(left.compose(&right).mat(None)),
        &cmtm.compose_mat,
        1e-10,
    );

    approx_eq_matrix_dynamic(
        &dmatrix_to_vec(mathroborust::RustCmtm::hat(&cmtm.derivatives)),
        &cmtm.hat,
        1e-10,
    );
    approx_eq_matrix_dynamic(
        &dmatrix_to_vec(mathroborust::RustCmtm::hat_adj(&cmtm.derivatives)),
        &cmtm.hat_adj,
        1e-10,
    );
    approx_eq_matrix_dynamic(
        &dmatrix_to_vec(mathroborust::RustCmtm::hat_commute_adj(&cmtm.derivatives)),
        &cmtm.hat_commute_adj,
        1e-10,
    );
    approx_eq_vector_list(
        &mathroborust::RustCmtm::vee(&mathroborust::RustCmtm::hat(&cmtm.derivatives)),
        &cmtm.vee,
        1e-10,
    );
    approx_eq_vector_list(
        &mathroborust::RustCmtm::vee_adj(&mathroborust::RustCmtm::hat_adj(&cmtm.derivatives)),
        &cmtm.vee_adj,
        1e-10,
    );
    approx_eq_matrix_dynamic(
        &dmatrix_to_vec(left.tangent_mat(None)),
        &cmtm.tangent_mat,
        1e-10,
    );
    approx_eq_matrix_dynamic(
        &dmatrix_to_vec(left.tangent_mat_cm(None)),
        &cmtm.tangent_mat_cm,
        1e-10,
    );
}
